ykcapi/src/lib.rs

```diff
@@ -91,19 +91,18 @@ mod c_testing {
     }
 
     /// Fetches the function name (`res_func`) and the block index (`res_bb`) at position `idx` in
-    /// `trace`. The caller must free() the function name.
+    /// `trace`.
     #[no_mangle]
     pub extern "C" fn __yktrace_irtrace_get(
         trace: *mut IRTrace,
         idx: usize,
-        res_func: *mut *mut c_char,
+        res_func: *mut *const c_char,
         res_bb: *mut usize,
     ) {
         let trace = unsafe { Box::from_raw(trace) };
         let blk = trace.get(idx).unwrap();
-        let c_func = CString::new(blk.func_name()).unwrap();
         unsafe {
-            *res_func = c_func.into_raw();
+            *res_func = blk.func_name().as_ptr();
```

Comment:
`func_name` returns a `&CStr`, whose pointer you then assign to `res_func`. Isn't the string still owned by `blk`, which goes out of scope at the end of this function. Or is `trace` where `blk` comes from still sticking around? It is hard to see from this code that the `CStr` pointer is still valid after `__yktrace_irtrace_get` returns.

Reply:
The string is indeed owned by `blk` who in turn is owned by the (borrowed) `trace`. Thus the string lives on.